pkg/genericapiserver/enablement_config.go

```diff
@@ -0,0 +1,157 @@
+}
+
+// Specifies the overrides for various API group versions.
+// This can be used to enable/disable entire group versions or specific resources.
+type APIGroupVersionEnablementConfig struct {
+	// Whether to enable or disable this entire group version.
+	Enable bool
+
+	// DisabledResources lists the resources that are specifically disabled for a group/version
+	// DisabledResources trumps EnabledResources
+	DisabledResources sets.String
+
+	// EnabledResources lists the resources that should be enabled by default.  This is a little
+	// unusual, but we need it for compatibility with old code for now.  An empty set means
+	// enable all.
+	EnabledResources sets.String
+}
+
+// AnyVersion is used during resource enablement checks to see if any version of the resource is enabled
+const AnyVersion = "."
+
+type APIResourceEnablementConfig struct {
+	GroupVersionDisablers map[unversioned.GroupVersion]*APIGroupVersionEnablementConfig
+}
+
+func NewAPIResourceEnablementConfig() *APIResourceEnablementConfig {
+	return &APIResourceEnablementConfig{GroupVersionDisablers: map[unversioned.GroupVersion]*APIGroupVersionEnablementConfig{}}
+}
+
+func NewAPIGroupVersionEnablementConfig() *APIGroupVersionEnablementConfig {
+	return &APIGroupVersionEnablementConfig{Enable: true, DisabledResources: sets.String{}, EnabledResources: sets.String{}}
+}
+
+func (o *APIResourceEnablementConfig) DisableVersions(versions ...unversioned.GroupVersion) {
+	for _, version := range versions {
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].Enable = false
+	}
+}
+
+func (o *APIResourceEnablementConfig) EnableVersions(versions ...unversioned.GroupVersion) {
+	for _, version := range versions {
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].Enable = true
+	}
+}
+
+func (o *APIResourceEnablementConfig) DisableResources(resources ...unversioned.GroupVersionResource) {
+	for _, resource := range resources {
+		version := resource.GroupVersion()
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].EnabledResources.Delete(resource.Resource)
+		o.GroupVersionDisablers[version].DisabledResources.Insert(resource.Resource)
+	}
+}
+
+func (o *APIResourceEnablementConfig) EnableResources(resources ...unversioned.GroupVersionResource) {
+	for _, resource := range resources {
+		version := resource.GroupVersion()
+		_, versionExists := o.GroupVersionDisablers[version]
+		if !versionExists {
+			o.GroupVersionDisablers[version] = NewAPIGroupVersionEnablementConfig()
+		}
+
+		o.GroupVersionDisablers[version].EnabledResources.Insert(resource.Resource)
```

Comment:
> same here, inserting to the list could make something that previously allowed all resources suddenly allow only a particular resource?

It's unusual, but I think until we refactor further, if you choose to start enabling a per-resource level, you've decided to manage them all yourself.